coap-gateway/bootstrapper/main.go

```diff
@@ -0,0 +1,222 @@
+package main
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"strings"
+
+	"github.com/google/uuid"
+	cmV1 "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	cmMetaV1 "github.com/jetstack/cert-manager/pkg/apis/meta/v1"
+	certManager "github.com/jetstack/cert-manager/pkg/client/clientset/versioned"
+	cmInformers "github.com/jetstack/cert-manager/pkg/client/informers/externalversions"
+	cfgLoader "github.com/plgd-dev/cloud/pkg/config"
+	"github.com/plgd-dev/kit/security/generateCertificate"
+	coreV1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/cache"
+)
+
+type Configuration struct {
+	CloudID           string                            `yaml:"cloudID"`
+	SecretName        string                            `yaml:"secretName"`
+	IssuerName        string                            `yaml:"issuerName"`
+	CertFileName      string                            `yaml:"certFileName"`
+	KeyFileName       string                            `yaml:"keyFileName"`
+	CAFileName        string                            `yaml:"caFileName"  `
+	IncludeCA         bool                              `yaml:"includeCA"`
+	CertConfiguration generateCertificate.Configuration `yaml:"certConfiguration"`
+}
+
+type Bootstrapper struct {
+	configuration     *Configuration
+	namespace         string
+	kubernetesClient  *kubernetes.Clientset
+	certManagerClient *certManager.Clientset
+}
+
+func main() {
+	var bs Bootstrapper
+	bs.loadConfiguration()
+	bs.initializeClients()
+	csr, pkeyBlock := bs.generateCSR()
+
+	certRequest := cmV1.CertificateRequest{
+		ObjectMeta: metaV1.ObjectMeta{
+			Name:      bs.configuration.SecretName,
+			Namespace: bs.namespace,
+		},
+		Spec: cmV1.CertificateRequestSpec{
+			Request: csr,
+			IsCA:    false,
+			IssuerRef: cmMetaV1.ObjectReference{
+				Name: bs.configuration.IssuerName,
+			},
+		},
+	}
+
+	bs.cleanupCertificateRequest()
+	factory := cmInformers.NewSharedInformerFactory(bs.certManagerClient, 0)
+	informer := factory.Certmanager().V1().CertificateRequests().Informer()
+	informerStopCh := make(chan struct{})
+
+	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			certRequest := obj.(*cmV1.CertificateRequest)
+			if bs.isCertificateReady(certRequest) {
+				bs.createSecret(pkeyBlock, certRequest.Status.Certificate, certRequest.Status.CA)
```

Comment:
pass `ctx` as first argument of `bs.createSecret = `bs.createSecret(ctx,....)`

Reply:
done

Reply:
Not resovlved it is much better when you pass ctx as argument.